ReactWindows/ReactNative.Shared/Bridge/NativeModuleWrapperBase.cs

```diff
@@ -0,0 +1,217 @@
+using System;
+using System.Collections.Generic;
+
+namespace ReactNative.Bridge
+{
+    /// <summary>
+    /// A wrapper class for native modules codegen.
+    /// </summary>
+    /// <typeparam name="T">Type of native module.</typeparam>
+    public abstract class NativeModuleWrapperBase<T> : INativeModule, INativeModuleWrapper
+        where T : INativeModule
+    {
+        /// <summary>
+        /// Instantiates the <see cref="NativeModuleWrapperBase{T}"/>.
+        /// </summary>
+        /// <param name="module">The native module being wrapped.</param>
+        public NativeModuleWrapperBase(T module)
+        {
+            Module = module;
+        }
+
+        /// <summary>
+        /// The methods callable from JavaScript on this module.
+        /// </summary>
+        public abstract IReadOnlyDictionary<string, INativeMethod> Methods { get; }
+
+        /// <summary>
+        /// The action queue used by the native module.
+        /// </summary>
+        /// <remarks>
+        /// Can be <code>null</code>, in which case, the call is evaluated
+        /// inline on the native modules action queue.
+        /// </remarks>
+        public IActionQueue ActionQueue => Module.ActionQueue;
+
+        /// <summary>
+        /// Return true if you intend to override some other native module that
+        /// was registered, e.g., as part of a different package (such as the
+        /// core one). Trying to override without returning true from this 
+        /// method is considered an error and will throw an exception during
+        /// initialization. By default, all modules return false.
+        /// </summary>
+        public bool CanOverrideExistingModule => Module.CanOverrideExistingModule;
+
+        /// <summary>
+        /// The constants exported by this module.
+        /// </summary>
+        public JObject Constants => Module.Constants;
+
+        /// <summary>
+        /// The native module instance.
+        /// </summary>
+        public T Module { get; }
+
+        /// <summary>
+        /// The native module type.
+        /// </summary>
+        public Type ModuleType => typeof(T);
+
+        /// <summary>
+        /// The name of the module.
+        /// </summary>
+        public string Name => Module.Name;
+
+        /// <summary>
+        /// Called after the creation of a <see cref="IReactInstance"/>, in
+        /// order to initialize native modules that require the React or
+        /// JavaScript modules.
+        /// </summary>
+        public void Initialize() => Module.Initialize();
+
+        /// <summary>
+        /// Called before a <see cref="IReactInstance"/> is disposed.
+        /// </summary>
+        public void OnReactInstanceDispose() => Module.OnReactInstanceDispose();
+
+        INativeModule INativeModuleWrapper.Module => Module;
+
+        /// <summary>
+        /// Wrapper native method implementation.
+        /// </summary>
+        protected class NativeMethod : INativeMethod
```

Comment:
can we split these nested classes across files by making it a partial class?

Reply:
I don't see the point of that.

Reply:
The intent is that the file is generated, I can't think of a scenario where you'd want to to modify this file or any of the classes in it.

Reply:
the other day I was told to split a delegate into a separate file to avoid multiple types per file, so just going for consistency. I agree that if it's generated, it matters even less.